core/src/main/scala/org/apache/spark/status/AppStatusStore.scala

```diff
@@ -222,29 +223,20 @@ private[spark] class AppStatusStore(
     val indices = quantiles.map { q => math.min((q * count).toLong, count - 1) }
 
     def scanTasks(index: String)(fn: TaskDataWrapper => Long): IndexedSeq[Double] = {
-      Utils.tryWithResource(
-        store.view(classOf[TaskDataWrapper])
-          .parent(stageKey)
-          .index(index)
-          .first(0L)
-          .closeableIterator()
-      ) { it =>
-        var last = Double.NaN
-        var currentIdx = -1L
-        indices.map { idx =>
-          if (idx == currentIdx) {
-            last
-          } else {
-            val diff = idx - currentIdx
-            currentIdx = idx
-            if (it.skip(diff - 1)) {
-              last = fn(it.next()).toDouble
-              last
-            } else {
-              Double.NaN
-            }
-          }
-        }.toIndexedSeq
+      val quantileTasks = store.view(classOf[TaskDataWrapper])
+        .parent(stageKey)
+        .index(index)
+        .first(0L)
+        .asScala
+        .filter(_.status.startsWith("S")) // Filter "SUCCESS" tasks
+        .zipWithIndex
+        .filter(x => indices.contains(x._2))
+
+      if(quantileTasks.size > indices.length) {
+          quantileTasks.map(task => fn(task._1).toDouble).toIndexedSeq
+      } else {
+        indices.map( index =>
+         fn(quantileTasks.filter(_._2 == index).head._1).toDouble).toIndexedSeq
```

Comment:
.filter + .head = .find? it is more efficient. Why not create a Map though if this needs to be accessed by key?

Reply:
Modified. Thanks